src/lib/shim/src/tls.rs

```diff
@@ -0,0 +1,610 @@
+    f: F,
+    _phantom: PhantomData<T>,
+}
+// SAFETY: Still `Sync` even if T is `!Sync`, since each thread gets its own
+// instance of the value. `F` must still be `Sync`, though, since that *is*
+// shared across threads.
+unsafe impl<T, F> Sync for ShimTlsVar<T, F> where F: Sync + Fn() -> T {}
+
+impl<T, F> ShimTlsVar<T, F>
+where
+    F: Fn() -> T,
+{
+    /// Create a variable that will be uniquely instantiated for each thread.
+    /// Typically this should go in a `static`.
+    pub const fn new(f: F) -> Self {
+        Self {
+            offset: LazyLock::const_new(OffsetInitializer::new::<T>()),
+            f,
+            _phantom: PhantomData,
+        }
+    }
+
+    const fn align() -> usize {
+        let align = core::mem::align_of::<ShimTlsVarStorage<T>>();
+        // Must be at least the alignment of `ShimThreadLocalStorage`.
+        // If this fails, we can increase `ShimThreadLocalStorage`'s alignment.
+        assert!(align <= 16);
+        align
+    }
+
+    fn offset(&self) -> usize {
+        *self.offset.force()
+    }
+
+    /// Access the inner value.
+    ///
+    /// The returned wrapper can't be sent to or shared with other threads, to
+    /// allow the implementation to be extended in the future to deallocate
+    /// and/or reuse storage when a thread exits.
+    pub fn get(&self) -> TLSVarRef<T> {
+        // SAFETY: We ensured `offset` is in bounds at construction time.
+        let this_var_bytes: *mut u8 = {
+            // We need a mut pointer to the bytes, but can't create a mut
+            // reference to the whole storage object, since that may alias other
+            // active variables in this thread.
+            //
+            // We can't use a const pointer to the storage as a whole to get to
+            // the _bytes field, because we don't have another layer of
+            // UnsafeCell to opt into being allowed to get another mut reference
+            // from a const pointer.
+            //
+            // Instead we rely on the fact that the storage object is `repr[C]`
+            // and has `_bytes` as the first field, so the pointer to `_bytes`
+            // has the same value as the pointer to the enclosing object.
+            let storage: *mut u8 = tls_storage().get() as *mut u8;
+            unsafe { storage.add(self.offset()) }
+        };
+
+        let this_var: &ShimTlsVarStorage<T> = {
+            let this_var: *const ShimTlsVarStorage<T> =
+                this_var_bytes as *const ShimTlsVarStorage<T>;
+            assert_eq!(this_var.align_offset(Self::align()), 0);
+            // SAFETY: The TLS bytes for each thread are initialized to 0, and
+            // all-zeroes is a valid value of `ShimTlsVarStorage<T>`.
+            //
+            // We've ensure proper alignment when calculating the offset,
+            // and verified in the assertion just above.
+            let this_var: &ShimTlsVarStorage<T> = unsafe { &*this_var };
+            this_var
+        };
```

Comment:
I think most of this code would be nicer as a method of `ShimTlsVarStorage` instead. It's essentially casting an `&UnsafeCell<ShimThreadLocalStorage>` with an offset to a `&ShimTlsVarStorage<T>`, and I think this would be better as a self-contained function that explains exactly why this is safe.